commons-db/src/main/java/com/palantir/nexus/db/sql/SQLString.java

```diff
@@ -251,22 +252,49 @@ private static String makeCommentString(String keyString, DBType dbType) {
     /**
      * Cleans up whitespace, any trailing semicolon, and prefixed comments that a string is
      * unregistered, in order to come up with a canonical representation of this sql string.
-     * @param s
-     * @return
      */
-    public static String canonicalizeString(String s) {
-        String trim = s.trim();
-        if(!trim.isEmpty() && trim.charAt(trim.length() - 1) == ';') {
-            trim = trim.substring(0, trim.length() - 1);
+    public static String canonicalizeString(String sql) {
+        return canonicalizeString(sql, false);
+    }
+
+    static String canonicalizeStringAndRemoveWhitespaceEntirely(String sql) {
+        return canonicalizeString(sql, true);
+    }
+
+    private static String canonicalizeString(String original, boolean removeAllWhitespaceEntirely) {
+        StringBuilder cleanedString = new StringBuilder(original);
+        int originalIdx = 0;
+        int cleanedIdx = 0;
+        int firstUnregisteredIdx = cleanedString.indexOf(UNREGISTERED_SQL_COMMENT);
+
+        while (originalIdx < original.length()) {
+            char originalChar = original.charAt(originalIdx);
+            if (originalIdx == firstUnregisteredIdx) {
+                originalIdx += UNREGISTERED_SQL_COMMENT.length();
+                firstUnregisteredIdx = original.indexOf(UNREGISTERED_SQL_COMMENT, originalIdx);
+            } else if (originalChar == ' ' && (cleanedIdx == 0 || cleanedString.charAt(cleanedIdx - 1) == ' ')
+                    || (originalChar == ' ' && removeAllWhitespaceEntirely)) {
```

Comment:
The above looks like `(A and (B or C) or (A and D))`; is there a reason we didn't write it as `(A and (B or C or D))`?